### PR TITLE
New Filtering System for Text Search

### PR DESCRIPTION
**Pull request by firenoo**
Changes how the search bar works and how results are sorted.

- Searching by tag: You can now filter entries by "tags", which are separated words starting with `@`. Currently 4 tags supported:
    - `@in` and `@out` filters by input/output P2Ps, respectively
    - `@b` and `@u` filters by bound and unbound P2Ps, respectively
    - Use single/double quotes to search for the literal string (see below)
- Searching by name: Instead of searching for the literal string, each space-separated, non-tag word will be its own query. To force searching for a literal string, you can use quotes (single or double)

Results are now sorted by:
1. Selected frequency at the top (input before outputs)
2. Most number of words matching the search queries
    - Further sorted by shortest name -> longest  name
3. The rest of the frequencies if no search queries exist

### CODE REVIEW

**Comment by firenoo** [comment]
Slated for post 2.3

File: src/main/java/com/projecturanus/betterp2p/client/gui/InfoFilter.kt [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    val activeFilters: MutableMap<Filter, MutableList<String>?> = mutableMapOf()
+
+    /**
+     * Parse the query string for filters and update the active
+     * filter list.
+     */
+    fun updateFilter(query: String) {
+        // I spent 10 minutes on this until I gave up... regex wtf
+        // https://stackoverflow.com/questions/366202/regex-for-splitting-a-string-using-space-when-not-surrounded-by-single-or-double
+        val regex = "[^\\s\"']+|\"([^\"]*)\"|'([^']*)'".toRegex()
```
Comment (eigenraven) [inline src/main/java/com/projecturanus/betterp2p/client/gui/InfoFilter.kt:26]: Keep this regex as a class property, so it's not recompiled on every use.

File: src/main/java/com/projecturanus/betterp2p/client/gui/InfoFilter.kt [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+ * - By bound/unbound: Use `@b` or `@u` filters by input/output respectively.
+ * - By name: Use the name
+ * If someone comes and says "sort by freq pls" then we can add it at that time
+ */
+class InfoFilter {
+
+    /**
+     * Active filters to use when filtering entries.
+     */
+    val activeFilters: MutableMap<Filter, MutableList<String>?> = mutableMapOf()
```
Comment (eigenraven) [inline src/main/java/com/projecturanus/betterp2p/client/gui/InfoFilter.kt:17]: Afaik you can't put nulls as values into maps, even in Kotlin. Nulls and missing values are treated the same, use empty lists or Optionals instead.

File: src/main/java/com/projecturanus/betterp2p/client/gui/InfoFilter.kt [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        }
+    }
+}
+
+/**
+ * The different filter types. Probably will let these be adjustable in the
+ * config.
+ */
+enum class Filter(val pattern: Regex, val filter: (InfoWrapper, List<String>?) -> Boolean) {
+    INPUT("\\A@in\\z".toRegex(), { it, _ -> !it.output }),
```
Comment (eigenraven) [inline src/main/java/com/projecturanus/betterp2p/client/gui/InfoFilter.kt:62]: You should display the syntax somewhere, for example in a tooltip if you hover over the entry textbox

File: src/main/java/com/projecturanus/betterp2p/client/gui/InfoList.kt [hunk 4]
```diff
@@ -69,12 +72,42 @@ class InfoList (initList: Collection<InfoWrapper>,
...
+                            hits += 1
+                            name = name.replaceFirst(f, "", true)
+                        }
+                    }
+                    -(hits * hits) + name.length
+                }
+                else -> it.frequency + Short.MAX_VALUE
+            }
         }
+//        filtered = sorted.filter {
```
Comment (eigenraven) [inline src/main/java/com/projecturanus/betterp2p/client/gui/InfoList.kt:105]: Remove commented code

File: src/main/java/com/projecturanus/betterp2p/client/gui/InfoFilter.kt (hunk 2 above)
Comment (firenoo) [reply src/main/java/com/projecturanus/betterp2p/client/gui/InfoFilter.kt:17]: They should: [see this playground](https://play.kotlinlang.org/#eyJ2ZXJzaW9uIjoiMS44LjEwIiwicGxhdGZvcm0iOiJqYXZhIiwiYXJncyI6IiIsIm5vbmVNYXJrZXJzIjp0cnVlLCJ0aGVtZSI6ImlkZWEiLCJjb2RlIjoiZnVuIG1haW4oKSB7XG5cblx0dmFsIG1hcDogTXV0YWJsZU1hcDxTdHJpbmcsIFN0cmluZz8+ID0gbXV0YWJsZU1hcE9mKClcblxuICAgIG1hcFtcIlRlc3RcIl0gPSBudWxsXG4gICAgbWFwW1wiSGVsbG9cIl0gPSBudWxsXG4gICBcdHByaW50bG4obWFwW1wiVGVzdFwiXSlcbiAgICBwcmludGxuKG1hcFtcIkhlbGxvXCJdKVxuICAgIHByaW50bG4obWFwLmtleXMpXG59In0=)

The intended use is that the other filters' values are don't cares.

**Review by eigenraven — APPROVED** [review]

**Pull request merged**